chia/full_node/mempool_manager.py

```diff
@@ -425,18 +388,35 @@ async def validate_spend_bundle(
         log.debug(f"Cost: {cost}")
 
         assert npc_result.conds is not None
-        # build set of removals
-        removal_names: Set[bytes32] = set(bytes32(spend.coin_id) for spend in npc_result.conds.spends)
-        if removal_names != set(s.name() for s in new_spend.removals()):
-            # If you reach here it's probably because your program reveal doesn't match the coin's puzzle hash
-            return Err.INVALID_SPEND_BUNDLE, None, []
-
-        additions: List[Coin] = additions_for_npc(npc_result)
+        removal_names: Set[bytes32] = set()
         additions_dict: Dict[bytes32, Coin] = {}
         addition_amount: int = 0
-        for add in additions:
-            additions_dict[add.name()] = add
-            addition_amount = addition_amount + add.amount
+        eligibility_and_additions: Dict[bytes32, Tuple[bool, List[Coin]]] = {}
+        non_eligible_coin_ids: List[bytes32] = []
+        for spend in npc_result.conds.spends:
+            coin_id = bytes32(spend.coin_id)
+            removal_names.add(coin_id)
+            spend_additions = []
+            for puzzle_hash, amount, _ in spend.create_coin:
+                child_coin = Coin(coin_id, puzzle_hash, amount)
+                spend_additions.append(child_coin)
+                additions_dict[child_coin.name()] = child_coin
+                addition_amount = addition_amount + child_coin.amount
+            is_eligible = bool(spend.flags & ELIGIBLE_FOR_DEDUP)
+            if not is_eligible:
+                non_eligible_coin_ids.append(coin_id)
+            eligibility_and_additions[coin_id] = (is_eligible, spend_additions)
+        removal_names_from_coin_spends: Set[bytes32] = set()
+        bundle_coin_spends: Dict[bytes32, BundleCoinSpend] = {}
+        for coin_spend in new_spend.coin_spends:
+            coin_id = coin_spend.coin.name()
+            removal_names_from_coin_spends.add(coin_id)
+            eligible_for_dedup, spend_additions = eligibility_and_additions.get(coin_id, (False, []))
+            bundle_coin_spends[coin_id] = BundleCoinSpend(coin_spend, eligible_for_dedup, spend_additions)
```

Comment:
it looks like this new logic, added to `validate_spend_bundle()` does not have a test. unless I'm missing it.